Add unit tests for runtime.py

bgeltz opened the pull request:
- Resolves #1803.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add unit tests for runtime.py

- Resolves #1803.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo commented on service/geopmdpy/runtime.py:
```diff
@@ -85,10 +85,17 @@ def __init__(self, period, num_period=None):
 
             num_period (int): Number of time periods spanned by the
-                              loop.
+                              loop + 1.
```
We should clarify in the documentation.  The total loop time is num_periods * period, but since there is no delay in the first loop, there will be num_period + 1 loop iterations.

cmcantalupo replied:
```suggestion
                              loop.  The total loop time is num_periods * period,
                              but since there is no delay in the first loop, there
                              will be num_period + 1 loop iterations.
```

cmcantalupo commented on service/geopmdpy/runtime.py:
```diff
@@ -238,13 +247,16 @@ def run(self):
             pid = subprocess.Popen(self._argv)
             for loop_idx in TimedLoop(self._update_period, self._num_update):
                 if pid.poll() is not None:
+                    # Nice process has died error, and RC
```
Totally, we should be tracking the return code of the subprocess and add an accessor method to the Controller object so it can be returned from main (should update the test_controller.py accordingly).

cmcantalupo commented on service/geopmdpy/runtime.py:
```diff
@@ -238,13 +247,16 @@ def run(self):
                 signals = self.read_all_signals()
                 controls = self._agent.update(signals)
                 for control_idx in self._controls_idx:
-                    pio.adjust(control_idx, control[control_idx])
```
Nice!

cmcantalupo commented on service/geopmdpy_test/TestRuntime.py:
```diff
@@ -0,0 +1,237 @@
+    def get_report(self):
+        return self._report_data
+
+class TestRuntime(unittest.TestCase):
```
This works, but I'd rather have the test classes map to the classes they are testing.  It means we have to have multiple test files for runtime.py, TestTimedLoop.py and TestController.py, and the tests for TimedLoop and Controller would be independent.

cmcantalupo commented on service/geopmdpy_test/TestRuntime.py:
```diff
@@ -0,0 +1,237 @@
+        with self.assertRaisesRegex(StopIteration, ''):
+            next(it)
+
+    def test_agent(self):
```
Could probably leave this class in the TestController test class in TestController.py.  Seems a little overkill to have a separate file to test a class without an implementation.  May be worth checking that all methods have a __doc__ string that is not empty.

cmcantalupo commented on service/geopmdpy_test/TestRuntime.py:
```diff
@@ -0,0 +1,237 @@
+
+        # Again, with timeout, same expectations otherwise
+        timeout = 10
+        with mock.patch('geopmdpy.pio.push_signal', side_effect = itertools.count()) as pps, \
```
what is the itertools.count() doing here?

bgeltz replied:
https://docs.python.org/3/library/itertools.html#itertools.count

I'm using it so that every time you call the mocked push_signal, it'll return an ever increasing integer, starting at 0 in this case.  Here's an example:

```
>>> import itertools
>>> from unittest import mock
>>> m = mock.MagicMock()
>>> m()
<MagicMock name='mock()' id='140437089806936'>
>>> m.side_effect = itertools.count()
>>> m()
0
>>> m()
1
>>> m()
2
```

bgeltz committed:
Add negative tests

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Refactor

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add more negative tests

bgeltz committed:
Add returncode api

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Reload modules that may have been previously mocked

- Resolves issue that was exposed with TestController.py
  where mocks from TestController.py were being invoked by
  tests that run later because of alphabetical ordering
  when executing through pytest.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add subprocess error test

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add new tests to Makefile

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo committed:
Clarify the documentation for TimedLoop constructor

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
